test/rpc_test.py

```diff
@@ -508,6 +508,44 @@ def test_shutdown(self):
         # it's safe to call shutdown() multiple times
         rpc.shutdown()
 
+    @dist_init(clean_shutdown=False)
+    def test_wait_all_workers(self):
+        # worker0 drives and waits for worker1 and worker2
+        # throughout the test.
+        if self.rank == 0:
+            assert self.world_size >= 3
+
+            num_repeat = 30
+            futs = []
+
+            # Phase 1: Only worker1 has workload.
+            dst = "worker1"
+            for _ in range(num_repeat):
+                fut = rpc.rpc_async(dst, heavy_rpc, args=(torch.ones(100, 100),))
+                futs.append(fut)
+
+            for fut in futs:
+                fut.wait()
+                self.assertEqual(fut.wait(), 0)
+
+            # Phase 2: Only worker2 has workload.
+            # If join is not correctly implemented,
+            # worker2 should be closed by now.
+            dst = "worker2"
+            for _ in range(num_repeat):
+                fut = rpc.rpc_async(dst, heavy_rpc, args=(torch.ones(100, 100),))
+                futs.append(fut)
+
+            for fut in futs:
+                fut.wait()
+                self.assertEqual(fut.wait(), 0)
+
+        # worker0 calls this at the end after waiting for RPC responses.
+        # worker1/2 calls this immediately and has some works after it.
+        # worker3 calls this immediately and has no more work.
+        rpc.api._wait_all_workers()
+        rpc.shutdown(graceful=False)
```

Comment:
So this test will actually call `shutdown` twice with `graceful=False`. This will be a little complicated to convey to users: allowing calling `shutdown` arbitrary times as long as `graceful=False` but only once if `graceful=True`. Is this intentional?

Reply:
I think the call to shutdown is redundant, I wanted to call `_wait_all_workers()` explicitly to make sure it is truly tested.

Not intentional for shutdown. I think it's the conclusion that we want to first make wait_all_workers only support being called for once.

I can make it be ing supported to be called multiple times. But that will introduce a state, like context id.

Reply:
>  I think it's the conclusion that we want to first make wait_all_workers only support being called for once.

Yes, and I think that's good enough for now. But, we do allow calling `shutdown` multiple times, right?

Reply:
This change essentially make shutdown does not support calling multiple times. An exception would be thrown on the second call.